pkg/experiment/metastore/raftnode/snapshot.go

```diff
@@ -0,0 +1,89 @@
+package raftnode
+
+import (
+	"fmt"
+	"io/fs"
+	"os"
+	"path/filepath"
+
+	"github.com/go-kit/log/level"
+)
+
+// importSnapshots moves snapshots from SnapshotsImportDir to SnapshotsDir.
+// It skips snapshot files that already exist in the destination directory, and
+// moves the rest.
+//
+// When initializing the snapshot store, Raft creates a "snapshots" subdirectory.
+// To maintain consistency, this behavior is replicated here – so "snapshots" should
+// not be included in the configured path.
+//
+// Note: the import process does not guarantee atomicity, as it may involve moving
+// files across different file systems. If an error occurs during the import,
+// both the source and destination directories may be left in an inconsistent state.
+// The function does not attempt to recover from such a state, but it will try to
+// continue copying on the next call.
+func (n *Node) importSnapshots() error {
+	if n.config.SnapshotsImportDir == "" {
+		return nil
+	}
+
+	importDir := filepath.Join(n.config.SnapshotsImportDir, "snapshots")
+	snapshotsDir := filepath.Join(n.config.SnapshotsDir, "snapshots")
+
+	level.Info(n.logger).Log("msg", "importing snapshots")
+	entries, err := fs.ReadDir(os.DirFS(importDir), ".")
+	if err != nil {
+		return fmt.Errorf("failed to read dir: %w", err)
+	}
+
+	for _, entry := range entries {
+		if !entry.IsDir() {
+			continue
+		}
+		dstDir := filepath.Join(snapshotsDir, entry.Name())
+		srcDir := filepath.Join(importDir, entry.Name())
+		if err = n.moveSnapshot(srcDir, dstDir); err != nil {
+			return fmt.Errorf("failed to import snapshot %q: %w", entry.Name(), err)
+		}
+	}
+
+	return nil
+}
+
+func (n *Node) moveSnapshot(srcDir, dstDir string) error {
+	entries, err := fs.ReadDir(os.DirFS(srcDir), ".")
+	if err != nil {
+		return fmt.Errorf("failed to read snapshot dir: %w", err)
+	}
+	var hasFiles bool
+	for _, entry := range entries {
+		if entry.IsDir() {
+			continue
+		}
+		hasFiles = true
+		break
+	}
+	if !hasFiles {
+		return nil
+	}
+
+	level.Info(n.logger).Log("msg", "importing snapshot", "snapshot", srcDir)
+	if err = os.MkdirAll(dstDir, 0755); err != nil {
+		return fmt.Errorf("failed to create snapshot directory: %w", err)
+	}
+	for _, entry := range entries {
+		srcPath := filepath.Join(srcDir, entry.Name())
+		dstPath := filepath.Join(dstDir, entry.Name())
+		if _, err = os.Stat(dstPath); err == nil {
+			// While move is not guaranteed to be atomic, we do expect
+			// that if a file is found, it is a valid snapshot file.
+			level.Info(n.logger).Log("msg", "file already exists, skipping", "file", entry.Name())
+			continue
+		}
+		if err = os.Rename(srcPath, dstPath); err != nil {
```

Comment:
The comment above says it will be moving across different filesystems.
If I remember correctly rename syscall does not work across different file systems.

Reply:
That's a good point – I overlooked quite a few things there. Fixed!